### ISSUE TITLE
restarting a kubelet should never affect the running workload

### ISSUE THREAD

**Issue by davshen** [kubernetes/kubernetes#123980]
### What happened?

1.Node label changed.
Node label changed is because the operation and maintenance engineer organized the node label, such as hpc=true, and removed this label. When a pod is compatible with this label, restarting the node kubelet will cause the pod to be rebuilt. This actually shouldn’t be the case. It will affect the normal operation of the business.
2.https://github.com/kubernetes/kubernetes/issues/123971
3.https://github.com/kubernetes/kubernetes/issues/123816
After the above three scenarios occur, as long as the kubelet is restarted, the pod will be evicted. Obviously, this is not as expected. It is also an undesirable result for online business.

### What did you expect to happen?

Restarting a kubelet should not cause any disruption to the running workload (which likely will mean skip admission of running pods, but let's not run ahead of ourselves) and backlink to this issue is probably the best way forward still.

### How can we reproduce it (as minimally and precisely as possible)?

Before restarting the kubelet, fully check the scenarios mentioned above.

### Kubernetes version

<details>

```
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.17", GitCommit:"a7736eaf34d823d7652415337ac0ad06db9167fc", GitTreeState:"clean", BuildDate:"2022-12-08T11:47:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.17", GitCommit:"a7736eaf34d823d7652415337ac0ad06db9167fc", GitTreeState:"clean", BuildDate:"2022-12-08T11:42:04Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by davshen** [comment]
https://github.com/kubernetes/kubernetes/issues/123859

**Comment by davshen** [comment]
I think when the admit fails, the pod creation time and current status should be judged. If the pod is currently in the running state, and the pod creation time is earlier than the kubelet startup time, skip rejecting the pod.

**Comment by ffromani** [comment]
/triage accepted
/priority important-longterm

We are aware of cases on which a kubelet restart perturbs the workload, and it shouldn't. We will collect the cases on this issue.
I'm torn between `backlog` and `important-longterm`. Setting the latter because otherwise this issue will likely slip through cracks.

**Comment by davshen** [comment]
https://github.com/kubernetes/kubernetes/issues/118559 Here is another case about "Running pods with devices are terminated if kubelet is restarted"
The way he modified it was https://github.com/kubernetes/kubernetes/pull/119432

**Comment by Mihawk-lf** [comment]
How long does it take for kubelet to restart?

**Comment by davshen** [comment]
> How long does it take for kubelet to restart?
> 
> My test is normal. If the kubelet restart takes too long, it will cause the pod to be evicted. The pod fixed on this machine will be restarted.


You need to label the node first, such as hpc=true, and then create a pod that is compatible with this label. After the creation is successful, remove the hpc=true label and restart kubelet, and a pod failed error will appear.

**Comment by pacoxu** [comment]
I prefer to think this is a designed behavior.

If the node label is changed and the pod affinity doesn't match, the pod can be rejected during kubelet restart. 

/cc @yujuhong @dchen1107

**Comment by yujuhong** [comment]
I agree that it's worth revisiting the pod admission logic in kubelet, but the particular case of rejecting pods due to changed labels seems working as intended. I don't think I fully understand the use case of this. Maybe someone can help elaborate more?

**Comment by Homura222** [comment]
> I prefer to think this is a designed behavior.
> 
> If the node label is changed and the pod affinity doesn't match, the pod can be rejected during kubelet restart.(especially when node label is changed by kubelet flags.)
> 
> ```
> --node-labels <key=value pairs>
> <Warning: Alpha feature>Labels to add when registering the node in the cluster. Labels must be key=value pairs separated by ','. Labels in the 'kubernetes.io' namespace must begin with an allowed prefix ('kubelet.kubernetes.io', 'node.kubernetes.io') or be in the specifically allowed set ('beta.kubernetes.io/arch', 'beta.kubernetes.io/instance-type', 'beta.kubernetes.io/os', 'failure-domain.beta.kubernetes.io/region', 'failure-domain.beta.kubernetes.io/zone', 'kubernetes.io/arch', 'kubernetes.io/hostname', 'kubernetes.io/os', 'node.kubernetes.io/instance-type', 'topology.kubernetes.io/region', 'topology.kubernetes.io/zone')
> ```
> 
> /cc @yujuhong @dchen1107

I believe that if this behavior is by design, then the pod should be evicted when the node label is changed and the pod affinity doesn't match, rather than waiting until the kubelet restarts to reject the pod.

**Comment by gjkim42** [comment]
> I don't think I fully understand the use case of this. Maybe someone can help elaborate more?

+1

---

The kubelet should not check any node affinity or pod affinity, as they are all about the scheduler.

refs:
- https://github.com/kubernetes/kubernetes/pull/101218#discussion_r627495203
- https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

> Note: In the preceding types, IgnoredDuringExecution means that if the node labels change after Kubernetes schedules the Pod, the Pod continues to run.

**Comment by chengjoey** [comment]
> I agree that it's worth revisiting the pod admission logic in kubelet, but the particular case of rejecting pods due to changed labels seems working as intended. I don't think I fully understand the use case of this. Maybe someone can help elaborate more?

For example, some pods have affinity set, `kubernetes.io/region` in `Asia`. At this time, these pods are running normally on the asia nodes. Then remove the label `kubernetes.io/region` of these asia nodes. After removing the label, the pods running on them are still running normally. However, after restarting the kubelet on these nodes, the pods running normally are killed and new nodes that meet the requirements are selected.

Either kubelet should kill these pods after asia nodes removes the label, or let them continue to run instead of rescheduling them after kubelet restart.

**Comment by gjkim42** [comment]
> However, after restarting the kubelet on these nodes, the pods running normally are killed and new nodes that meet the requirements are selected.

I think this is the bug. kubelet should let them continue to run after its restart, as node affinity should be considered only in the kube-scheduler.

**Comment by aojea** [comment]
This has a lot of edge cases with a very large blast, last time we did a small change on the lifecycle of the Pods it took several releases to stabilize again , I think this requires a KEP to deeply discuss the scope as per @ffromani comment here https://github.com/kubernetes/kubernetes/issues/123980#issuecomment-2064061961 and, if accepted, rollout this change with a feature gate so we have the capacity to evaluate the impact in production and roll back if necessary

**Comment by ffromani** [comment]
I agree with @aojea  that a resolution would totally need a KEP and close oversight. It's still very worthy to discuss the issue, which has its merit. Perhaps is time to further split this issue and differentiate between the labelling issue (I tend to agree with @gjkim42) and the exclusive resource allocation issue.

**Comment by davshen** [comment]
> I agree that it's worth revisiting the pod admission logic in kubelet, but the particular case of rejecting pods due to changed labels seems working as intended. I don't think I fully understand the use case of this. Maybe someone can help elaborate more?

For example, I have many different types of nodes. At first, they were labeled with many types of labels, including the label hpc=true. However, after a long time, the cluster administrator wanted to sort out the labels of all nodes in the cluster. I mistakenly deleted the hpc=true label. At this time, if you restart the kubelet, the pod with hpc=true affinity will be evicted and no suitable node can be found, causing an online business failure.

**Comment by davshen** [comment]
> > I agree that it's worth revisiting the pod admission logic in kubelet, but the particular case of rejecting pods due to changed labels seems working as intended. I don't think I fully understand the use case of this. Maybe someone can help elaborate more?
> 
> For example, some pods have affinity set, `kubernetes.io/region` in `Asia`. At this time, these pods are running normally on the asia nodes. Then remove the label `kubernetes.io/region` of these asia nodes. After removing the label, the pods running on them are still running normally. However, after restarting the kubelet on these nodes, the pods running normally are killed and new nodes that meet the requirements are selected.
> 
> Either kubelet should kill these pods after asia nodes removes the label, or let them continue to run instead of rescheduling them after kubelet restart.

我同意

**Comment by aojea** [comment]
but the system should not compensate for human mistakes ... what if he change was one purpose and the admin wanted the pod to be evicted?

**Comment by davshen** [comment]
> > I agree that it's worth revisiting the pod admission logic in kubelet, but the particular case of rejecting pods due to changed labels seems working as intended. I don't think I fully understand the use case of this. Maybe someone can help elaborate more?
> 
> For example, some pods have affinity set, `kubernetes.io/region` in `Asia`. At this time, these pods are running normally on the asia nodes. Then remove the label `kubernetes.io/region` of these asia nodes. After removing the label, the pods running on them are still running normally. However, after restarting the kubelet on these nodes, the pods running normally are killed and new nodes that meet the requirements are selected.
> 
> Either kubelet should kill these pods after asia nodes removes the label, or let them continue to run instead of rescheduling them after kubelet restart.

@aojea Perhaps this behavior, I think, is more appropriate. The eviction action should not occur when the kubelet is restarted, but when the label is deleted. This is more in line with the final state-oriented design concept of k8s.

**Comment by gjkim42** [comment]
I think the issue of evicting pods based on node affinity should be separated from this issue.
It is a well-documented behavior and looks like an obvious bug to me. At least we have to update the document instead.

https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

![Screenshot 2024-04-23 at 3 51 20 PM](https://github.com/kubernetes/kubernetes/assets/19814047/5b15fcd0-3601-4b77-b863-6ca0970b107b)

**Comment by AllenXu93** [comment]
I think the point is that restart kubelet should not affect any running workload. 
I have met similar scenario, when restarting Kubelet, some running pod failed due to a init-container NUMA calculation error.
https://github.com/kubernetes/kubernetes/issues/123816#issuecomment-1995112194
Both this case or node's label change should not make pod failed.

**Comment by aojea** [comment]
> I think the point is that restart kubelet should not affect any running workload

what do we mean by restart, it is killing the kubelet process to start a new one, no? does "restart" has a duration or can be undetermined the time of a restart?
what if I stop kubelet , do some maintenance and start it again, is that considered a "restart"?
what if something has changed in the node that impact the workloads?

**Comment by AllenXu93** [comment]
> > I think the point is that restart kubelet should not affect any running workload
> 
> what do we mean by restart, it is killing the kubelet process to start a new one, no? does "restart" has a duration or can be undetermined the time of a restart? what if I stop kubelet , do some maintenance and start it again, is that considered a "restart"? what if something has changed in the node that impact the workloads?

You are right. Maybe we can add a feature gate, the operator should know what whether this restart need to evict pods.
As you said the first type restart `killing the kubelet process to start a new one`, operator can turn on feature gate to make sure his operation safety

**Comment by gjkim42** [comment]
> I think the issue of evicting pods based on node affinity should be separated from this issue. It is a well-documented behavior and looks like an obvious bug to me. At least we have to update the document instead.
> 
> https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity
> 
> ![Screenshot 2024-04-23 at 3 51 20 PM](https://private-user-images.githubusercontent.com/19814047/324725211-5b15fcd0-3601-4b77-b863-6ca0970b107b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyODExNTIsIm5iZiI6MTcxNDI4MDg1MiwicGF0aCI6Ii8xOTgxNDA0Ny8zMjQ3MjUyMTEtNWIxNWZjZDAtMzYwMS00Yjc3LWI4NjMtNmNhMDk3MGIxMDdiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI4VDA1MDczMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2MjhjNWEzZjMwYTk2YjU1YzdhMzJlZDlkOGI3NzA1ZDdkNzY5ZjQ3YmRhYTU2NDE4MGI5Y2EwYjkwYmI1ZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JY69UtGNRlSd5Jqel-TtK2f1Bl2hy_WqTHSh_CMrRd4)

opened an issue for this.

**Comment by rmohr** [comment]
> but the system should not compensate for human mistakes ...

@aojea I think this is an issue for long running operations. Examples would be: AI jobs
 
 With the current implementation basically the following is said: Every cluster admin has to ensure that pods are safely drained or done executing before the kubelet gets restarted if they plan to restart (and this can even happen happen involuntarily with NPD). So relabeling nodes would become a super-hard task and not about mistakes necessarily
 
 >  what if the change was done purpose and the admin wanted the pod to be evicted?
 
Relying here on a kubelet restart to make this happen does not sound like something which anyone is using? To actually properly do this, there would be `RequiredDuringSchedulingIgnoreDuringExecution` and `RequiredDuringSchedulingRequiredDuringExecution`?

**Comment by chenk008** [comment]
Add another case: https://github.com/kubernetes/kubernetes/issues/125397

**Comment by xuzhenglun** [comment]
Here is some my throught about why current behaviour is problemic:
1. current eviction behaviour does not care about availaibility machiasm, such as PDB. When we are upgrading kubelet, it's hard to detect if the evication is caused by upgradion, and we need to stop upgrading.
2. If Pod need to be evicated when label is not matched anymore, this should happen at once when label is changed, not when restart kubelet. This behaviour is very confusing and make problem hard to dignosis, just kinda like a landmines laid in the past are triggered in the present by a seemingly insignificant act.

**Comment by aojea** [comment]
I don't think there is a disagreement that there is a problem and we want it to get it fixed, I think that the question is how to implement it with the confidence we don't break any existing workload that depend on current behavior ...  can we narrow the problem down to something more specific?

**Comment by haircommander** [comment]
/triage accepted

still a bug, though I still think we don't have consensus on how to fix
